File: framework/web/response.py
from datetime import UTC, datetime
from typing import Generic, TypeVar

from pydantic import BaseModel, Field

from framework.observability.context import request_id_ctx

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    code: str = "OK"
    message: str = "success"
    data: T | None = None
    request_id: str = Field(default_factory=request_id_ctx.get)
    timestamp: datetime = Field(default_factory=lambda: datetime.now(UTC))


def ok(data: T | None = None, *, message: str = "success") -> ApiResponse[T]:
    return ApiResponse[T](data=data, message=message)


def failure(*, code: str, message: str, data: object | None = None) -> ApiResponse[object]:
    return ApiResponse[object](code=code, message=message, data=data)

File: apps/channel_gateway/src/response.ts
export interface ApiEnvelope<T> {
  code: string;
  message: string;
  data: T;
  request_id: string;
  timestamp: string;
}

export function ok<T>(data: T, requestId: string, message = 'success'): ApiEnvelope<T> {
  return {
    code: 'OK',
    message,
    data,
    request_id: requestId,
    timestamp: new Date().toISOString(),
  };
}

export function failure(
  code: string,
  message: string,
  requestId: string,
  data: unknown = null,
): ApiEnvelope<unknown> {
  return {
    code,
    message,
    data,
    request_id: requestId,
    timestamp: new Date().toISOString(),
  };
}
